File: migrations/src/test/java/aga/android/migrations/SchemaHelperTest.java
package aga.android.migrations;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;

import static java.util.Collections.singletonList;
import static org.junit.Assert.assertEquals;

public class SchemaHelperTest {

    private static final String PARENT_TABLE_DEFINITION = "create table Parent (" +
    " id integer primary key not null," +
    " name varchar(255) default \'John Doe\'" +
    ");";

    private static final String CHILDREN_TABLE_DEFINITION = "create table Child (" +
    " id integer primary key not null," +
    " name text," +
    " parentId integer not null," +
    " foreign key (parentId) references Parent (id)" +
    ");";

    @Test
    public void testColumnInfoExtraction() {
        final Schema expected = new Schema.SchemaBuilder()
            .setTables(
                singletonList(
                    new Table.Builder()
                        .setName("Parent")
                        .setColumns(
                            Arrays.asList(
                                new Column.Builder()
                                    .setName("id")
                                    .setType(Affinity.INTEGER)
                                    .setIsNullable(false)
                                    .setIsPrimaryKey(true)
                                    .createColumn(),
                                new Column.Builder()
                                    .setName("name")
                                    .setType(Affinity.TEXT)
                                    .setIsNullable(true)
                                    .createColumn()
                            )
                        )
                        .createTable()
                )
            )
            .createSchema();

        final Schema actual = SchemaHelper.toSchema(singletonList(PARENT_TABLE_DEFINITION));

        assertEquals(expected, actual);
    }

    @Test
    public void testForeignKeyInfoExtraction() {
        final Schema expected = new Schema.SchemaBuilder()
            .setTables(
                singletonList(
                    new Table.Builder()
                        .setName("Child")
                        .setColumns(
                            Arrays.asList(
                                new Column.Builder()
                                    .setName("id")
                                    .setType(Affinity.INTEGER)
                                    .setIsNullable(false)
                                    .setIsPrimaryKey(true)
                                    .createColumn(),
                                new Column.Builder()
                                    .setName("name")
                                    .setType(Affinity.TEXT)
                                    .setIsNullable(true)
                                    .createColumn(),
                                new Column.Builder()
                                    .setName("parentId")
                                    .setType(Affinity.INTEGER)
                                    .setIsNullable(false)
                                    .createColumn()
                            )
                        )
                        .setForeignKeys(
                            Collections.singletonList(
                                new ForeignKey.Builder()
                                    .setKeyName("parentId")
                                    .setReferenceTable("Parent")
                                    .setReferenceField("id")
                                    .createForeignKey()
                            )
                        )
                        .createTable()
                )
            )
            .createSchema();

        final Schema actual = SchemaHelper.toSchema(singletonList(CHILDREN_TABLE_DEFINITION));

        assertEquals(expected, actual);
    }
}

File: migrations/src/test/java/aga/android/migrations/SchemaHelperTest.kt
package aga.android.migrations

import aga.android.migrations.SchemaHelper.toSchema
import org.junit.Assert.assertEquals
import org.junit.Test

class SchemaHelperTest {
    @Test
    fun testColumnInfoExtraction() {
        val expected = Schema(
            setOf(
                Table(
                    name = "Parent",
                    columns = (
                        setOf(
                            Column.Builder()
                                .setName("id")
                                .setType(Affinity.INTEGER)
                                .setIsNullable(false)
                                .setIsPrimaryKey(true)
                                .build(),
                            Column.Builder()
                                .setName("name")
                                .setDefaultValue("\'John Doe\'")
                                .setType(Affinity.TEXT)
                                .setIsNullable(true)
                                .build()
                        )
                    )
                )
            )
        )

        val actual = toSchema(listOf(PARENT_TABLE_DEFINITION))
        assertEquals(expected, actual)
    }

    @Test
    fun testForeignKeyInfoExtraction() {
        val expected = Schema(
            tables = setOf(
                Table(
                    name = "Child",
                    columns = setOf(
                        Column.Builder()
                            .setName("id")
                            .setType(Affinity.INTEGER)
                            .setIsNullable(false)
                            .setIsPrimaryKey(true)
                            .build(),
                        Column.Builder()
                            .setName("name")
                            .setType(Affinity.TEXT)
                            .setIsNullable(true)
                            .build(),
                        Column.Builder()
                            .setName("parentId")
                            .setType(Affinity.INTEGER)
                            .setIsNullable(false)
                            .build()
                        ),
                    foreignKeys = setOf(
                        ForeignKey.Builder()
                            .setKeyName("parentId")
                            .setReferenceTable("Parent")
                            .setReferenceField("id")
                            .setOnDelete(ForeignKeyAction.NO_ACTION)
                            .setOnUpdate(ForeignKeyAction.CASCADE)
                            .build()
                    )
                )
            )
        )

        val actual = toSchema(listOf(CHILDREN_TABLE_DEFINITION))
        assertEquals(expected, actual)
    }

    @Test
    fun testLowerAndUpperCaseKeyWordsDoNoAffectSchema() {
        val lowerCaseSchema = toSchema(listOf(PARENT_TABLE_DEFINITION))
        val upperCaseSchema = toSchema(listOf(PARENT_TABLE_DEFINITION_CAPS))
        assertEquals(lowerCaseSchema, upperCaseSchema)
    }

    @Test
    fun testKeywordOrderDoesNotAffectSchema() {
        val schema1 = toSchema(listOf(PARENT_TABLE_DEFINITION))
        val schema2 = toSchema(listOf(PARENT_TABLE_DEFINITION_DIFFERENT_ORDER))
        assertEquals(schema1, schema2)
    }

    companion object {

        private const val PARENT_TABLE_DEFINITION = "create table Parent (" +
            " id integer primary key not null," +
            " name varchar(255) default \'John Doe\'" +
            ");"

        private const val PARENT_TABLE_DEFINITION_DIFFERENT_ORDER = "create table Parent (" +
            " id integer not null primary key," +
            " name varchar(255) default \'John Doe\'" +
            ");"

        private const val PARENT_TABLE_DEFINITION_CAPS = "CREATE TABLE Parent (" +
            " id INTEGER PRIMARY KEY NOT NULL," +
            " name VARCHAR(255) DEFAULT \'John Doe\'" +
            ");"

        private const val CHILDREN_TABLE_DEFINITION = "create table Child (" +
            " id integer primary key not null," +
            " name text," +
            " parentId integer not null," +
            " foreign key (parentId) references Parent (id) on update cascade on delete no action" +
            ");"
    }
}